### PR TITLE
KAFKA-6042: Avoid deadlock between two groups with delayed operations

### PR DESCRIPTION
**Pull request by rajinisivaram**

### CODE REVIEW

**Comment by rajinisivaram** [comment]
@ijuma @hachikuji Can you review please? I was trying to avoid changing group/txnMetadata locking last time,  but that was a mistake. This goes back to the previous way of overriding locks, but with `ReentrantLock`.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: nit: move this to previous line?

**Review by hachikuji — APPROVED** [review]
Left a nit, but LGTM overall. It might help to find somewhere to add some notes on the locking strategy both for the `GroupCoordinator` and the `TransactionCoordinator`.

**Comment by hachikuji** [comment]
@rajinisivaram Hmm.. Do we need to update `DelayedHeartbeat` also to use the group lock as the operation lock.

**Comment by rajinisivaram** [comment]
@hachikuji I updated `DelayedHeartbeat` to use group lock and `DelayedTxnMarker` to use state lock. Also added comments in `GroupCoordinator` and `TransactionStateManager`. Can you take another look? Thank you!

**Comment by hachikuji** [comment]
The updates look good to me. I'll wait for @guozhangwang to take a second look before merging.

**Review by guozhangwang — COMMENTED** [review]
I have a meta question for my own understanding: the original reported deadlock in KAFKA-6042 seems between two offset commit request for two groups, one holding group A and tries to unblock for the other parked delayed producer for group B, and the other vice versa. This PR seems to be focusing on the transaction metadata manager instead. Does this actually fix the reported issue?

Otherwise I do not have any comments. Nice replacing `synchronized` with lock.

**Comment by rajinisivaram** [comment]
@guozhangwang The PR fixes deadlocks in both group coordinator and transaction manager. For groups, the group lock is now used as the delayed operation lock . The deadlock reported was exactly as you described. Now, when offset commit  request for groupA tries to complete groupB, the completion is performed if and only if it is able to acquire groupB lock (using `tryLock`). And similarly for groupB completing groupA. So we may acquire two group locks in any order, but the cycle is broken by `tryLock`, preventing deadlock.

**Comment by rajinisivaram** [comment]
please retest this

**Comment by rajinisivaram** [comment]
retest this please

**Comment by guozhangwang** [comment]
> And similarly for groupB completing groupA. So we may acquire two group locks in any order, but the cycle is broken by tryLock, preventing deadlock.

Thanks for the explanation. LGTM.

**Comment by guozhangwang** [comment]
The jenkins failures seem un-related. I'm going to merge to trunk and 1.0.

**Comment by guozhangwang** [comment]
Merged to 1.0, and cherry-picked to trunk.

File: core/src/main/scala/kafka/server/DelayedOperation.scala
Comment (ijuma) [inline core/src/main/scala/kafka/server/DelayedOperation.scala:47]: I suggest reconsidering all default arguments used for locks. Since it's crucial to get them right, it's generally safer to force callers to choose. My general recommendation is to avoid default arguments whenever correctness is involved.

**Comment by boweite** [comment]
@ijuma @hachikuji @guozhangwang Hi, I wonder if this patch can completely solve this problem. Because I found the order of the parameters passed from replicaManager.appendRecords is wrong. So the group.lock is not passed to delayedProduce, so there is also the risk of deadlock. Please check it.

**Pull request closed without merging**
